python/lsst/ip/diffim/imageMapReduce.py

```diff
@@ -248,38 +250,76 @@ def run(self, mapperResults, exposure, **kwargs):
             subExp = afwImage.ExposureF(newExp, item.getBBox())
             subMI = subExp.getMaskedImage()
             patchMI = item.getMaskedImage()
-            isNan = (np.isnan(patchMI.getImage().getArray()) |
-                     np.isnan(patchMI.getVariance().getArray()))
-            noNans = np.sum(isNan) <= 0
+            isNotNan = ~(np.isnan(patchMI.getImage().getArray()) |
+                         np.isnan(patchMI.getVariance().getArray()))
             if reduceOp == 'copy':
-                if noNans:
-                    subMI[:, :] = patchMI
-                else:
-                    isNotNan = ~isNan
-                    subMI.getImage().getArray()[isNotNan] = patchMI.getImage().getArray()[isNotNan]
-                    subMI.getVariance().getArray()[isNotNan] = patchMI.getVariance().getArray()[isNotNan]
+                subMI.getImage().getArray()[isNotNan] = patchMI.getImage().getArray()[isNotNan]
+                subMI.getVariance().getArray()[isNotNan] = patchMI.getVariance().getArray()[isNotNan]
 
             if reduceOp == 'sum' or reduceOp == 'average':  # much of these two options is the same
-                if noNans:
-                    subMI += patchMI
-                    if reduceOp == 'average':
-                        # wsubim is a view into the `weights` Image, so here we simply add one to
-                        # the region of `weights` confined by `item.getBBox()`.
-                        wsubim = afwImage.ImageF(weights, item.getBBox())
-                        wsubim += 1.
-                else:
-                    isNotNan = ~isNan
-                    subMI.getImage().getArray()[isNotNan] += patchMI.getImage().getArray()[isNotNan]
-                    subMI.getVariance().getArray()[isNotNan] += patchMI.getVariance().getArray()[isNotNan]
-                    if reduceOp == 'average':
-                        wsubim = afwImage.ImageF(weights, item.getBBox())
-                        wsubim.getArray()[isNotNan] += 1.
+                subMI.getImage().getArray()[isNotNan] += patchMI.getImage().getArray()[isNotNan]
+                subMI.getVariance().getArray()[isNotNan] += patchMI.getVariance().getArray()[isNotNan]
+                if reduceOp == 'average':
+                    # wsubim is a view into the `weights` Image
+                    wsubim = afwImage.ImageI(weights, item.getBBox())
+                    wsubim.getArray()[isNotNan] += 1
 
         if reduceOp == 'average':
-            newMI /= weights
+            wts = weights.getArray().astype(np.float)
```

Comment:
Good catch, I'm already working on setting a mask value to invalid in this case. This will be implemented as part of DM-7611 (about to submit for review).